Fix several supervisor issues and bugs

UncleGrumpy opened the pull request:
These changes fix several open supervisor issues, as well as a few small bugs discovered in the supervisor and test_supervisor.erl.

* Implements missing `one_for_all` restart strategy that is documented in the module.
* Supervisors now obeys `intensity` and `period` options.
* Children who fail to restart are retried until maximum restart intensity limit is reached
* Fixes edge case bug when terminating as well as several typos.
* Fixes some bugs in the test suite that left un-received messages in the test environment that would be received by additional tests added to the end of the lists in `test_supervisor:/test/0`.

Closes #1855
Closes #1915
Closed #1957

These changes are made under both the "Apache 2.0" and the "GNU Lesser General
Public License 2.1 or later" license terms (dual license).

SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later

UncleGrumpy committed:
Fix typos in supervisor

Fixes a typo in the init funtion error return where the atom `mod` was returned rather than the
modules name (`Mod`).

Fixes a typo in child record `type`, adds missing parenthesis on the `child_type/0` type.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Fix some bugs in test_supervisor.erl

Fixes some tests to catch the messages from child ping_pong_servers so they will not be received by
later tests. Changes the supervisor started in test_count_children to use a generic supervisor
without a child, rather than using the modules start_link which always starts a child
ping_pong_server that sends messages back to the test environment. Fixes the test_ping_ping final
`recieve` to match on a new `Pid4` (which would indicate a restart) rather than `Pid3` that the
monitor just caught the 'DOWN' message for.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Add support for supervisor one_for_all strategy

Adds support for handling restarts for the `one_for_all` strategy that was documented, but lacked
implementation. Makes necessary changes to ensure children are always restarted in the same order
they were originally started, and shutdown in reverse order with last child first, conforming to
OTP behavior.

Closes #1855

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Add support for supervisor intensity and period

Adds support for honoring intensity and period supervisor options, allowing the prevention of
endless crash and restarts by misbehaving children.

Adds a test to ensure the settings work as expected on AtomVM and match OTP.

Closes #1915

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Fix supervisor restart when child fails to match OTP

Fixes a function clause exception that would crash the supervisor if a child fails to restart by
matching the same behavior as OTP, and continue to try restarting the child until success, or
maximum `intensity` is reached within the allowed `period`.

Closed #1957

Signed-off-by: Winford <winford@object.stream>

petermm commented on tests/libs/estdlib/test_supervisor.erl:
```diff
@@ -340,3 +444,216 @@ test_which_children() ->
+                {error, no_supervisor2}
+        end,
+    ok =
+        case is_pid(ChildPid) of
```
believe you want ChildPid2 here

UncleGrumpy replied:
Yes I do! Thank you. I reworked this function a bit too, I decided it was a waist of CI time to wait 2000 ms for a child message if the supervisor failed to start 🤦 . I moved the check for a supervisor pid sooner and a fixed the variable name.

pguyot commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -309,6 +386,26 @@ handle_info({ensure_killed, Pid}, State) ->
             {noreply, State}
     end;
+handle_info({try_again_restart, Id}, State) ->
+    Child = lists:keyfind(Id, #child.id, State#state.children),
```
The child can be gone here.

pguyot commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -321,9 +418,16 @@ terminate(_Reason, #state{children = Children} = State) ->
 loop_terminate([#child{pid = Pid} = Child | Tail], AccRemaining) when is_pid(Pid) ->
-    do_terminate(Child),
-    loop_terminate(Tail, [Pid | AccRemaining]);
+    case is_process_alive(Pid) of
```
I'm not sure about this optimization. The process could be gone between here and the exit message, and it's not worth it unless we need to ensure it's gone.

pguyot reviewed (commented):
Thank you for this work!

I looked at OTP's implementation and we're doing things differently, which may or may not be a good thing :)

OTP's terminate_child/2 is synchronous and I like the asynchronous implementation we currently have. Yet, it introduces some complexity.

I wonder if we should move further and make this entirely state based, i.e. avoid `send_after/3` entirely and rely on timeout feature of gen_server instead for the general case, or just plain receive of exit messages for the terminate case. It may make this more easy to test or more pure. Also be careful of processing of messages with send_after because we don't know if the child is still there.

UncleGrumpy commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -184,7 +210,50 @@ start_children([Child | T], StartedC) ->
     StartedC.
 
-restart_child(Pid, Reason, State) ->
+restart_child(Pid, Reason, #state{restart_strategy = one_for_one} = State) ->
```
Yes, I found the name confusing too, since terminate/2 was added to main. I think making it clear that this is the handler for child exits is a very good idea.

UncleGrumpy committed:
Rename supervisor:restart_child/3 to remove confusion

Renames restart_child/3 to handle_child_exit/3 for clarity. This is an intenal callback handler,
and not related to the exported restart_child/2.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Supervisor cleanup: keep internal funtions together

Moves the internal private funtions `handle_child_exit/3` and `should_restart/2` to the same
section as the rest of the internal private funtions.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -321,9 +418,16 @@ terminate(_Reason, #state{children = Children} = State) ->
 
 loop_terminate([#child{pid = undefined} | Tail], AccRemaining) ->
     loop_terminate(Tail, AccRemaining);
+loop_terminate([#child{pid = {restarting, _}} | Tail], AccRemaining) ->
```
> we might want that for internal tracking when children are using timeout shutdown.

... actually this wouldn't make sense either, we are already tracking this in the `#state.restart`.

UncleGrumpy commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -201,13 +270,21 @@ restart_child(Pid, Reason, State) ->
+                                {ok, NewChildren} ->
+                                    {ok, State1#state{children = NewChildren}};
+                                {ok, NewChildren, RetryID} ->
+                                    erlang:send_after(50, self(), {try_again_restart, RetryID}),
```
Sadly, our current gen_server implementation does not support `{ok, State, {timeout, Time, Message}` returns from callbacks. I was thinking that adding hibernation support might be nice (especially for the benefit of supervisors), but I think supporting `timeout` actions in returns is even more important.

UncleGrumpy committed:
Update and add missing supervisor types, specs and module doc

- Updates some types to match OTP, and adds some others, including exports types for
`startchild_ret/0` and `startlink_ret/0`.
- Adds a module doc section listing differences with OTP supervisor.
- Adds specs to all public functions, and marks callbacks as hidden functions so they are not
included in published user API docs.
- Adds todos where child and supervisor event reports should be logged

Signed-off-by: Winford <winford@object.stream>

pguyot commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -70,65 +94,135 @@
+-type startchild_ret() ::
+    {ok, Child :: child()} | {ok, Child :: child(), Info :: term()} | {error, startchild_err()}.
+-type startchild_err() :: already_present | {already_started, Child :: child()} | term().
+-type sup_name() :: {local, Name :: atom()}.
```
OTP exports the `sup_name()`  type, so we should do it too.

pguyot commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -169,75 +263,44 @@ child_spec_to_record(#{id := ChildId, start := MFA} = ChildMap) ->
-                    {reply, Result, State#state{children = [UpdatedChild | Children]}};
+                    %% The last child to start should always be at the end of the child
+                    %% start list.
+                    {reply, Result, State#state{children = Children ++ [UpdatedChild]}};
```
Maybe we should keep the list reversed, i.e. append last item at the head in the most natural linked list way.

pguyot commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -364,6 +553,132 @@ try_start(#child{start = {M, F, Args}} = Record) ->
 
+terminate_one_for_all_restart(Children) ->
+    %% Always shut down last child first
+    do_terminate_one_for_all_restart(lists:reverse(Children), [], []).
```
We would terminate last child first if we stored the list in the natural order.

pguyot commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -364,6 +553,132 @@ try_start(#child{start = {M, F, Args}} = Record) ->
+    case Child of
+        #child{restart = {terminating, temporary, From}} = Child when is_pid(From) ->
+            do_terminate(Child),
+            case verify_shutdown(Child) of
```
I would queue all children for verification by default and avoid the 100ms wait here. Then I would loop until all exit message are received. I wonder about building a list of children to timeout kill instead of using the timers, but the timers can make a cleaner loop:

```erlang
terminate_loop([]) ->
    ok;
terminate_loop(Children) ->
    receive
        {'EXIT', Pid, _Reason} ->
            terminate_loop(lists:keydelete(Pid, #child.pid, Children));
        {ensure_killed, Pid} ->
            exit(Pid, kill),
            terminate_loop(Children)
    end.
```

UncleGrumpy commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -70,65 +94,135 @@
     modules = [] :: [module()] | dynamic
 }).
--record(state, {restart_strategy :: strategy(), children = [] :: [#child{}]}).
+%% note: the list of children should always be kept in order, with first to start at the head.
```
We do need to keep some specified order to ensure shutdown is in reverse order as startup, and restart always happen in the same order as startup.

I did have some concerns about the cost of additions, and list reversals after iterations with accumulators, but the accum reversals are often un-avoidable depending on the operation. I think this boils down to do we optimize restart time, so crashed children come back up as soon as possible, or do we optimize for startup and shutdown times at the expense of slower restarts?

UncleGrumpy replied:
You are absolutely correct about the order. Not sure what I was thinking, the second time I needed to use a lists:reverse/1 I should have realized this mistake. After I fixed this I realized I was not even improving restart speeds, it was penalizing performance all the way around.

UncleGrumpy commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -309,18 +374,142 @@ handle_info({ensure_killed, Pid}, State) ->
 terminate(_Reason, #state{children = Children} = State) ->
-    RemainingChildren = loop_terminate(Children, []),
+    %% Shutdown children last to first.
+    RemainingChildren = loop_terminate(lists:reverse(Children), []),
```
The shutdown order is specific that it is the reverse of start order.

See my comment above about whether we should optimize restarts or add child and shutdown operations.

UncleGrumpy commented on libs/estdlib/src/supervisor.erl:
```diff
@@ -70,65 +97,135 @@
+-type sup_ref() ::
+    (Name :: atom())
+    | {Name :: atom(), Node :: node()}
+    | {global, Name :: term()}
```
Good catch.
